### ISSUE TITLE
kbct mainthread panics/quits when separate bluetooth trackpad connects

### ISSUE THREAD

**Issue by seltzered** [samvel1024/kbct#6]
Steps:

1) Start machine (bluetooth keyboard on - paired but not yet connected, bluetooth trackpad on - paired but not yet connected)
2. login using bluetooth keyboard, haven't touched trackpad at all yet
3. launch terminal to run kbct
4. click trackpad a couple times causing it to connect

get similar error to #4 :
```
 thread 'main' panicked at 'Could not capture keyboard: IOError(Os { code: 2, kind: NotFound, message: "No such file or directory" })', src/main.rs:212:58
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

```
will try to run with backtrace to get better debug info.

**Comment by seltzered** [comment]
full stacktrace for trackpad connect crash scenario, essentially similar to what's been experienced in #4 

```
 INFO  kbct > Starting kbct event loop, pid=4080
 INFO  kbct > Capturing device path=/dev/input/event21 name="NUTYPE F1 Keyboard" mapped_name="Kbct-NUTYPE F1 Keyboard"
 
thread 'main' panicked at 'Could not capture keyboard: IOError(Os { code: 2, kind: NotFound, message: "No such file or directory" })', src/main.rs:212:58
stack backtrace:
   0:     0x56180b9beaf0 - std::backtrace_rs::backtrace::libunwind::trace::h5e9d00f0cdf4f57e
                               at /rustc/2fd73fabe469357a12c2c974c140f67e7cdd76d0/library/std/src/../../backtrace/src/backtrace/libunwind.rs:90:5
   1:     0x56180b9beaf0 - std::backtrace_rs::backtrace::trace_unsynchronized::hd5302bd66215dab9
                               at /rustc/2fd73fabe469357a12c2c974c140f67e7cdd76d0/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:     0x56180b9beaf0 - std::sys_common::backtrace::_print_fmt::ha0237cd11a34e2bf
                               at /rustc/2fd73fabe469357a12c2c974c140f67e7cdd76d0/library/std/src/sys_common/backtrace.rs:67:5
   3:     0x56180b9beaf0 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h171d4c10df1a98ee
                               at /rustc/2fd73fabe469357a12c2c974c140f67e7cdd76d0/library/std/src/sys_common/backtrace.rs:46:22
   4:     0x56180b9de00c - core::fmt::write::h89e4288724daa3fa
                               at /rustc/2fd73fabe469357a12c2c974c140f67e7cdd76d0/library/core/src/fmt/mod.rs:1096:17
   5:     0x56180b9bb852 - std::io::Write::write_fmt::h6d40f996e84584d9
                               at /rustc/2fd73fabe469357a12c2c974c140f67e7cdd76d0/library/std/src/io/mod.rs:1568:15
   6:     0x56180b9c0c35 - std::sys_common::backtrace::_print::h0c0b93221682afc8
                               at /rustc/2fd73fabe469357a12c2c974c140f67e7cdd76d0/library/std/src/sys_common/backtrace.rs:49:5
   7:     0x56180b9c0c35 - std::sys_common::backtrace::print::h57a9f95204c2fdd6
                               at /rustc/2fd73fabe469357a12c2c974c140f67e7cdd76d0/library/std/src/sys_common/backtrace.rs:36:9
   8:     0x56180b9c0c35 - std::panicking::default_hook::{{closure}}::h4245258b50e37e69
                               at /rustc/2fd73fabe469357a12c2c974c140f67e7cdd76d0/library/std/src/panicking.rs:208:50
   9:     0x56180b9c0793 - std::panicking::default_hook::h7b00dcc1d0944747
                               at /rustc/2fd73fabe469357a12c2c974c140f67e7cdd76d0/library/std/src/panicking.rs:225:9
  10:     0x56180b9c13d1 - std::panicking::rust_panic_with_hook::h71e6a073d87de1f5
                               at /rustc/2fd73fabe469357a12c2c974c140f67e7cdd76d0/library/std/src/panicking.rs:591:17
  11:     0x56180b9c0f17 - std::panicking::begin_panic_handler::{{closure}}::hd549436f6bb6dbb8
                               at /rustc/2fd73fabe469357a12c2c974c140f67e7cdd76d0/library/std/src/panicking.rs:497:13
  12:     0x56180b9bef8c - std::sys_common::backtrace::__rust_end_short_backtrace::h4e5f4b72b04174c3
                               at /rustc/2fd73fabe469357a12c2c974c140f67e7cdd76d0/library/std/src/sys_common/backtrace.rs:141:18
  13:     0x56180b9c0e79 - rust_begin_unwind
                               at /rustc/2fd73fabe469357a12c2c974c140f67e7cdd76d0/library/std/src/panicking.rs:493:5
  14:     0x56180b9dc9a1 - core::panicking::panic_fmt::hcd56f7f635f62c74
                               at /rustc/2fd73fabe469357a12c2c974c140f67e7cdd76d0/library/core/src/panicking.rs:92:14
  15:     0x56180b9dc7c3 - core::option::expect_none_failed::hf37eebedadde29e6
                               at /rustc/2fd73fabe469357a12c2c974c140f67e7cdd76d0/library/core/src/option.rs:1300:5
  16:     0x56180b57abec - core::result::Result<T,E>::expect::h6f4d50364a9eff26
                               at /home/sme/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/result.rs:997:23
  17:     0x56180b500966 - <kbct::DeviceManager as kbct::nio::EventObserver>::on_event::h19a1c27fd17ee60a
                               at /home/sme/src/kbct/src/main.rs:212:5
  18:     0x56180b4d2580 - kbct::nio::EventLoop::run::h81776f0a537dbce6
                               at /home/sme/src/kbct/src/nio.rs:46:11
  19:     0x56180b501ec6 - kbct::start_mapper::h4df6ea383a04ddea
                               at /home/sme/src/kbct/src/main.rs:287:2
  20:     0x56180b501680 - kbct::start_mapper_from_file_conf::h6abaeacc14181489
                               at /home/sme/src/kbct/src/main.rs:275:2
  21:     0x56180b502fd9 - kbct::main::hf140f9ed72c3bbb9
                               at /home/sme/src/kbct/src/main.rs:367:4
  22:     0x56180b5243d2 - core::ops::function::FnOnce::call_once::h437e19c21ddd9d80
                               at /home/sme/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ops/function.rs:227:5
  23:     0x56180b50f78a - std::sys_common::backtrace::__rust_begin_short_backtrace::h6ac56ed2bd5d58d5
                               at /home/sme/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/sys_common/backtrace.rs:125:18
  24:     0x56180b50bb56 - std::rt::lang_start::{{closure}}::ha608ed6b8114b7f3
                               at /home/sme/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/rt.rs:66:18
  25:     0x56180b9c17f7 - core::ops::function::impls::<impl core::ops::function::FnOnce<A> for &F>::call_once::h527fb2333ede305e
                               at /rustc/2fd73fabe469357a12c2c974c140f67e7cdd76d0/library/core/src/ops/function.rs:259:13
  26:     0x56180b9c17f7 - std::panicking::try::do_call::h309d8aee8149866c
                               at /rustc/2fd73fabe469357a12c2c974c140f67e7cdd76d0/library/std/src/panicking.rs:379:40
  27:     0x56180b9c17f7 - std::panicking::try::h75a60c31fd16bfc6
                               at /rustc/2fd73fabe469357a12c2c974c140f67e7cdd76d0/library/std/src/panicking.rs:343:19
  28:     0x56180b9c17f7 - std::panic::catch_unwind::h1f9892423e99bc00
                               at /rustc/2fd73fabe469357a12c2c974c140f67e7cdd76d0/library/std/src/panic.rs:431:14
  29:     0x56180b9c17f7 - std::rt::lang_start_internal::hd5b67df56ca01dae
                               at /rustc/2fd73fabe469357a12c2c974c140f67e7cdd76d0/library/std/src/rt.rs:51:25
  30:     0x56180b50bb27 - std::rt::lang_start::h324a23944f56788f
                               at /home/sme/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/rt.rs:65:5
  31:     0x56180b50478a - main
  32:     0x7f3a6b282565 - __libc_start_main
  33:     0x56180b4c00ae - _start
  34:                0x0 - <unknown>
```

**Comment by samvel1024** [comment]
Hi @seltzered . Thanks for reporting this. I think the two issues you have posted are related and have the same root cause. This happened to me as well and I don't think I have a good fix in mind. The only possible solution I have in mind is just swallowing the error and printing an error message instead of letting the process panic. Any further contribution is welcomed!

**Comment by seltzered** [comment]
Sounds like a reasonable plan for now, thanks for responding! At the moment just trying to see if can launch kbct on startup (on ubuntu so via systemd) and make sure it still runs.

**Comment by samvel1024** [comment]
@seltzered Fixed this issue with commit `81670e1479d34663063380b6832e938b75416fa5`

Tested for a week and it runs without errors.
